Add Scheduler and TokenBucket metrics

landonxjames opened the pull request:
*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

landonxjames committed:
First pass of Scheduler and TokenBucket metrics

aajtodd commented on aws-sdk-s3-transfer-manager/src/metrics/aggregators.rs:
```diff
@@ -165,46 +166,152 @@ impl TransferMetrics {
+    // Total tasks in flight
     inflight: Arc<AtomicU64>,
+    // High water mark for inflight tasks
+    max_inflight: Gauge,
```
Wouldn't a high water mark be a simple monotonic counter? A gauge is for instantaneous measurements usually

landonxjames replied:
Updated in https://github.com/awslabs/aws-s3-transfer-manager-rs/pull/125/commits/1d9e705f796e540b6281dfafed3279c0d606a291

aajtodd commented on aws-sdk-s3-transfer-manager/src/metrics/aggregators.rs:
```diff
@@ -165,46 +166,152 @@ impl TransferMetrics {
+    // High water mark for inflight tasks
+    max_inflight: Gauge,
+    // Count of permits acquired
+    permits_acquired: UpDownCounter,
```
Shouldn't this be the same as in-flight?

landonxjames replied:
I was thinking of them slightly differently. `inflight` was the futures in flight, and `permits_acquired` was the actual `OwnedPermit` structs handed out. But you're right those do functionally count the same thing and the time when they aren't aligned is tiny. Updated to only track inflight in https://github.com/awslabs/aws-s3-transfer-manager-rs/pull/125/commits/1d9e705f796e540b6281dfafed3279c0d606a291

aajtodd commented on aws-sdk-s3-transfer-manager/src/metrics/aggregators.rs:
```diff
@@ -165,46 +166,152 @@ impl TransferMetrics {
+pub(crate) struct TokenBucketMetrics {
+    max_tokens: u64,
+    available_tokens: UpDownCounter,
+    token_wait_time: Histogram,
```
I'm not sure if this will be meaningful or not given permit requests come with different token amounts requested.

landonxjames replied:
I wonder if tracking percentage of available tokens over time or something would be more useful?

aajtodd commented on aws-sdk-s3-transfer-manager/src/middleware/limit/concurrency/future.rs:
```diff
@@ -82,7 +82,13 @@ where
                         Ok(_permit) => {
                             let req = this.request.take().expect("request set");
                             let inflight = this.scheduler.metrics.increment_inflight();
+                            this.scheduler.metrics.update_max_inflight(inflight);
```
Shouldn't `increment_inflight()` just handle this?

landonxjames replied:
Moved inside `increment_inflight` in https://github.com/awslabs/aws-s3-transfer-manager-rs/pull/125/commits/1d9e705f796e540b6281dfafed3279c0d606a291

aajtodd commented on aws-sdk-s3-transfer-manager/src/runtime/token_bucket.rs:
```diff
@@ -149,32 +150,43 @@ pub(crate) struct TokenBucket {
     // NOTE: tokio semaphore is fair, permits are given out in the order requested
     semaphore: Arc<Semaphore>,
     mode: ConcurrencyMode,
+    tb_metrics: Arc<TokenBucketMetrics>,
```
I'm kind of wondering whether token bucket metrics provide value at this point in time. 🤔

landonxjames replied:
I am also not sure, but will probably leave them in place until I can actually run some benchmarks and test all of this on some real use cases. Can remove if we find it not necessary after that testing.

aajtodd replied:
Ok maybe add a comment or something to that effect

landonxjames committed:
refactor(metrics): streamline scheduler metrics architecture

Remove permit acquisition/release tracking and simplify inflight
request monitoring with consistent counter types.

ysaito1001 commented on aws-sdk-s3-transfer-manager/src/runtime/scheduler.rs:
```diff
@@ -21,30 +23,49 @@ use crate::types::ConcurrencyMode;
 pub(crate) struct Scheduler {
     token_bucket: TokenBucket,
-    pub(crate) metrics: SchedulerMetrics,
+    pub(crate) metrics: Arc<SchedulerMetrics>,
```
nit: will this end up double-`Arc` ing? Individual fields of `SchedulerMetrics` seem to have their types backed by `Arc` already, e.g. `Histogram`, `Gauge`?

aajtodd commented on aws-sdk-s3-transfer-manager/src/metrics/aggregators.rs:
```diff
@@ -165,46 +166,135 @@ impl TransferMetrics {
+    // High water mark for inflight tasks
+    max_inflight: IncreasingCounter,
+    // Count of failed permit acquisitions
+    permit_acquisition_failures: IncreasingCounter,
```
not sure we need this at this stage

aajtodd commented on aws-sdk-s3-transfer-manager/src/runtime/token_bucket.rs:
```diff
@@ -187,35 +199,69 @@ impl TokenBucket {
+    fn new(
+        sem: PollSemaphore,
+        tokens: u32,
+        tb_metrics: Arc<TokenBucketMetrics>,
```
this _feels_ off/wrong to need all these parameters for this future but I get why it's done. Wondering if there isn't a cleaner approach though.
